src/transaction_builder.js

```diff
@@ -164,43 +196,55 @@ function prepareInput (input, kpPubKey, redeemScript) {
     input.redeemScriptType = expanded.scriptType
     input.prevOutScript = input.prevOutScript || bscript.scriptHash.output.encode(redeemScriptHash)
     input.prevOutType = scriptTypes.P2SH
+    input.witness = false
 
   // maybe we have some prevOut knowledge
   } else if (input.prevOutType) {
-    // pay-to-scriptHash is not possible without a redeemScript
-    if (input.prevOutType === scriptTypes.P2SH) throw new Error('PrevOutScript is P2SH, missing redeemScript')
+    // embedded scripts are not possible without a redeemScript
+    if (input.prevOutType === scriptTypes.P2SH ||
+        input.prevOutType === scriptTypes.P2WSH) {
+      throw new Error('PrevOutScript is ' + input.prevOutType + ', requires redeemScript')
+    }
 
     // try to derive missing information using our kpPubKey
     expanded = expandOutput(input.prevOutScript, input.prevOutType, kpPubKey)
     if (!expanded.pubKeys) return
 
     input.pubKeys = expanded.pubKeys
     input.signatures = expanded.signatures
+    input.witness = (input.prevOutScript === scriptTypes.P2WPKH)
 
   // no prior knowledge, assume pubKeyHash
   } else {
     input.prevOutScript = bscript.pubKeyHash.output.encode(bcrypto.hash160(kpPubKey))
     input.prevOutType = scriptTypes.P2PKH
     input.pubKeys = [kpPubKey]
     input.signatures = [undefined]
+    input.witness = false
   }
 }
 
+var EMPTY_SCRIPT = new Buffer(0)
+
 function buildInput (input, allowIncomplete) {
   var signatures = input.signatures
   var scriptType = input.redeemScriptType || input.prevOutType
-  var scriptSig
+  var stack
 
   switch (scriptType) {
+    case scriptTypes.P2WPKH:
```

Comment:
The stack encoding of P2WPKH and P2PKH is the same, maybe do something like this? 
```
    case scripttypes.P2WPKH:
    case scriptTypes.P2PKH:
             ...
             break
```